### PR TITLE
[WIP] Separating Timer code from installation, fixes #4329

### LINKED ISSUE #4329: Separate Timer code from Timer Installation
Post #4261, we should split up the `Timer` code from the service installation, like we do with the other services. This will reduce our extension file sizes, since only the Runtime will install so only it needs the code.

We can also do this for `Platform`, maybe a few others. But, definitely `Timer` and `Platform`.

Re: https://github.com/ampproject/amphtml/pull/4261#issuecomment-236223083

### PR DESCRIPTION
**Pull request by AaronCritchley**
Still a WIP, working on fixing error with the retrieval of the service.

### CODE REVIEW

**Comment by erwinmombay** [comment]
hey @AaronCritchley good seeing you again!

/to @dvoytenko for review (feel free to redelegate)

File: src/service/timer-impl.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    return Promise.race([delayPromise, opt_racePromise]);
+  }
+}
+
+
+/**
+ * @param {!Window} window
+ */
+export function installTimerService(window) {
+  getService(window, 'timer', () => {
```
Comment (jridgewell) [inline src/service/timer-impl.js]: Please use the `fromClass` helper.

File: src/runtime.js [hunk 2]
```diff
@@ -74,6 +75,7 @@ export function installRuntimeServices(global) {
   installViewerService(global);
   installViewportService(global);
   installHistoryService(global);
+  installTimerService(global);
```
Comment (dvoytenko) [inline src/runtime.js]: It looks like timer should go before anything else. I see it's used by viewer and history services in particular.

File: src/service/timer-impl.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Requires polyfills in immediate side effect.
+import '../polyfills';
```
Comment (dvoytenko) [inline src/service/timer-impl.js]: Why is this?

/cc @erwinmombay

File: src/service/timer-impl.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    return Promise.race([delayPromise, opt_racePromise]);
+  }
+}
+
+
+/**
+ * @param {!Window} window
+ */
+export function installTimerService(window) {
+  fromClass(window, 'timer', Timer)
```
Comment (dvoytenko) [inline src/service/timer-impl.js]: `return` and `@return` above

File: src/service/timer-impl.js (hunk 3 above)
Comment (erwinmombay) [inline src/service/timer-impl.js]: looks like it was to fix something in closure compiler dependency ordering https://github.com/ampproject/amphtml/commit/084eac2f60e7e99da0b33f89a2517a0735686fe6

Comment (dvoytenko) [inline src/service/timer-impl.js]: Still needed?

**Comment by jridgewell** [comment]
Fixes https://github.com/ampproject/amphtml/issues/4329.

**Comment by jridgewell** [comment]
Looks like the tests are failing because it can't find the timer service. Probably need to register it somewhere. :sigh:

**Comment by AaronCritchley** [comment]
Yeah - I wasn't too sure where to register it embarrassingly. I tried a few different test suites but couldn't nail it down due to not being able to see where in the tests I was failing :-(

Any suggestions welcomed via here or Slack!

**Comment by AaronCritchley** [comment]
I signed it!

File: src/service/timer-impl.js (hunk 3 above)
Comment (erwinmombay) [inline src/service/timer-impl.js]: yes still needed

**Comment by jridgewell** [comment]
Looks like the issue is coming from [`Vsync`](https://github.com/ampproject/amphtml/blob/master/test/functional/test-vsync.js#L397-L401) tests. Moving that inside a `beforeEach` block will help you find the rest of the errors.

**Comment by erwinmombay** [comment]
wonder if its worthwhile adding new Error().stack wrapped in a localDev check for these service accessors?

**Comment by erwinmombay** [comment]
@AaronCritchley here's how i fixed the issue in test-vsync.js https://gist.github.com/erwinmombay/777f40eb04df69d173b35a4c89ec819f 

unfortunately since timer is one of the most core services out there there's a lot more files its affecting :(

to further debug, change `getExistingServiceForWindow` (src/service.js) to something like:

```js
export function getExistingServiceForWindow(win, id) {
  if (!win.services || !win.services[id]) {
    console.log(new Error().stack);
  }
  return dev().assert(
      win.services[id] && win.services[id].obj, `${id} service not found.`);
}
```

**Comment by erwinmombay** [comment]
@AaronCritchley i've submitted a fix so that you'll get a better stack trace if service isn't installed.

just need better accessor existence checks. so change the assert to:

```js
return dev().assert(win.services && win.services[id] && win.services[id].obj, `${id} service not found.`);
```

**Comment by dvoytenko** [comment]
@AaronCritchley @jridgewell @erwinmombay  Hmm. That's strange. All services should be installed for all tests via `_init_tests.js`, right?

**Comment by jridgewell** [comment]
Not when you're using a `fakeWin`.

**Comment by dvoytenko** [comment]
I see. Didn't notice fakewin there. Then that's the case - we need to install it manually. I really hope we'll eventually have a good fakeWin fixture.

**Comment by jridgewell** [comment]
Closed by #4961.

**Pull request closed without merging**